### PR TITLE
rustdoc: Move sidebar items into shared JavaScript.

### PR DESCRIPTION
**Pull request by lifthrasiir**
It had been a source of huge bloat in rustdoc outputs. Of course, we can simply disable compiler docs (as `rustc` generates over 90M of HTML) but this approach fares better even after such decision.

Each directory now has `sidebar-items.js`, which immediately calls `initSidebarItems` with a JSON sidebar data. This file is shared throughout every item in the sidebar. The current item is highlighted via a separate JS snippet (`window.sidebarCurrent`). The JS file is designed to be loaded asynchronously, as the sidebar is rendered before the content and slow sidebar loading blocks the entire rendering. For the minimal accessibility without JS, links to the parent items are left in HTML.

In the future, it might also be possible to integrate crates data with the same fashion: `sidebar-items.js` at the root path will do that. (Currently rustdoc skips writing JS in that case.)

This has a huge impact on the size of rustdoc outputs. Originally it was 326MB uncompressed (37.7MB gzipped, 6.1MB xz compressed); it is 169MB uncompressed (11.9MB gzipped, 5.9MB xz compressed) now. The sidebar JS only takes 10MB uncompressed & 0.3MB gzipped.

### CODE REVIEW

**Comment by lifthrasiir** [comment]
Maybe this can be another reason to pursue #21724. xz was able to exploit lots of duplications throughout the tarball well.

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -65,12 +65,10 @@ use html::item_type::ItemType;
 use html::layout;
 use html::markdown::Markdown;
 use html::markdown;
-use html::escape::Escape;
 use stability_summary;
 
 /// A pair of name and its optional document.
-#[derive(Clone, Eq, Ord, PartialEq, PartialOrd)]
-pub struct NameDoc(String, Option<String>);
+pub type NameDoc = (String, Option<String>);
```
Comment (huonw) [inline src/librustdoc/html/render.rs]: What's the reason for this change?

File: src/librustdoc/html/render.rs [hunk 2]
```diff
@@ -2211,9 +2207,11 @@ impl<'a> fmt::Display for Sidebar<'a> {
     fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
         let cx = self.cx;
         let it = self.item;
+        let parentlen = cx.current.len() - if it.is_mod() {1} else {0};
+
+        // this is not rendered via JS, as that would hamper the accessibility
```
Comment (huonw) [inline src/librustdoc/html/render.rs]: Hm, this comment doesn't quite have enough context to make sense.

File: src/librustdoc/html/render.rs (hunk 1 above)
Comment (lifthrasiir) [inline src/librustdoc/html/render.rs]: The auto-serialization of the struct is much longer than that of an anonymous tuple: `{"_field0":...,"_field1":...}` I think.

File: src/librustdoc/html/render.rs (hunk 2 above)
Comment (huonw) [inline src/librustdoc/html/render.rs]: Oh, maybe there could be a comment at the top of the block like

> The sidebar is designed to display items like sibling functions and modules, however to preserve space, we render that part via some shared javascript.

Comment (lifthrasiir) [inline src/librustdoc/html/render.rs]: The mention of accessibility is quite off, I admit. (My original intention is that, we should preserve as much HTML as possible unless that will result in bloat.) I'll reword that.

**Comment by alexcrichton** [comment]
@bors: r+ 4a6fb45ee1c51a67d75d99244446868b36f22cc6

I'm just blown away at how much low hanging fruit there is here, thanks so much @lifthrasiir! You are the hero of the download size of Rust :)

**Comment by arthurprs** [comment]
I will just leave this: Awesome :+1:

**Comment by steveklabnik** [comment]
:confetti_ball:

**Comment by Manishearth** [comment]
Needs https://github.com/Manishearth/rust/commit/205d9d8cce2dae01a02ff148386dae5c9f10f111

**Pull request merged**